include/seqan3/search/algorithm/detail/search.hpp

```diff
@@ -123,38 +125,58 @@ inline auto search_single(index_t const & index, query_t & query, configuration_
 
     // TODO: filter hits and only do it when necessary (depending on error types)
 
+    constexpr bool output_cursor = cfg_t::template exists<search_cfg::output<detail::search_output_index_cursor>>();
+
+    using result_t = std::conditional_t<output_cursor,
+                                        search_result_value_type<typename index_t::cursor_type>,
+                                        search_result_value_type<typename index_t::cursor_type, size_t, size_t>>;
+
+    std::vector<result_t> results;
+
     // output cursors or text_positions
-    if constexpr (cfg_t::template exists<search_cfg::output<detail::search_output_index_cursor>>())
+    if constexpr (output_cursor)
     {
-        return internal_hits;
+        for (auto && cursor : internal_hits)
+        {
+            results.emplace_back(0, cursor);
+        }
+
+        return results;
     }
     else
     {
-        using hit_t = std::conditional_t<index_t::text_layout_mode == text_layout::collection,
-                                         std::pair<typename index_t::size_type, typename index_t::size_type>,
-                                         typename index_t::size_type>;
-        std::vector<hit_t> hits;
-
         if constexpr (cfg_t::template exists<search_cfg::mode<detail::search_mode_best>>())
         {
             // only one cursor is reported but it might contain more than one text position
             if (!internal_hits.empty())
             {
-                auto text_pos = internal_hits[0].lazy_locate();
-                hits.push_back(text_pos[0]);
+                results.emplace_back(0, internal_hits[0], internal_hits[0].lazy_locate()[0]);
             }
         }
         else
         {
-            for (auto const & cur : internal_hits)
+            // Store the results of locate (either `size_t` or `std::pair<size_t, size_t>`) for erasing duplicates.
+            decltype(internal_hits[0].locate()) hits;
+
+            for (auto && cursor : internal_hits)
             {
-                for (auto const & text_pos : cur.locate())
+                for (auto const & text_pos : cursor.locate())
```

Comment:
position is an integral type? Then you can also capture it by value.

Reply:
It's `size_t` or `std::pair<size_t, size_t>`